include/openrave/openravecontainer.h

```diff
@@ -0,0 +1,265 @@
+// -*- coding: utf-8 -*-
+// Copyright (C) 2024 OpenRAVE
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/** \file openravecontainer.h
+    \brief Collection of high performance containers
+ */
+#ifndef OPENRAVE_CONTAINER_H
+#define OPENRAVE_CONTAINER_H
+
+#include <openrave/config.h>
+#include <openrave/openraveexception.h>
+
+#include <stdint.h>
+#include <vector>
+
+namespace OpenRAVE {
+
+namespace orcontainer {
+
+/// \brief Represents Key-value pair.
+template<typename DataType> struct NamedData
```

Comment:
How about name it to `NameIdData`?

Reply:
Actually, it seems vector of key-value pair is slower than having two separate vectors, for each of key and value. Let me restructure and remove this class.

Reply:
Really?! What's your test?

Reply:
```
g++ -O3 map_impl_test.cpp && ./a.out 
numElements=2
0 stdmap_insert::total=201.37 usec, average=2.0137 nsec(100000 samples)
1 ormap1_insert::total=703.282 usec, average=7.03282 nsec(100000 samples)
2 ormap2_insert::total=702.533 usec, average=7.02533 nsec(100000 samples)
3 stdmap_find::total=501.288 usec, average=5.01288 nsec(100000 samples)
4 ormap1_find::total=647.529 usec, average=6.47529 nsec(100000 samples)
5 ormap2_find::total=704.756 usec, average=7.04756 nsec(100000 samples)
numElements=8
6 stdmap_insert::total=2234.99 usec, average=22.3499 nsec(100000 samples)
7 ormap1_insert::total=3717.06 usec, average=37.1706 nsec(100000 samples)
8 ormap2_insert::total=3584.03 usec, average=35.8403 nsec(100000 samples)
9 stdmap_find::total=2756.14 usec, average=27.5614 nsec(100000 samples)
10 ormap1_find::total=3564.97 usec, average=35.6497 nsec(100000 samples)
11 ormap2_find::total=3819.25 usec, average=38.1925 nsec(100000 samples)
numElements=32
12 stdmap_insert::total=12598.2 usec, average=125.982 nsec(100000 samples)
13 ormap1_insert::total=38134.2 usec, average=381.342 nsec(100000 samples)
14 ormap2_insert::total=28884.3 usec, average=288.843 nsec(100000 samples)
15 stdmap_find::total=14316.1 usec, average=143.161 nsec(100000 samples)
16 ormap1_find::total=22810.9 usec, average=228.109 nsec(100000 samples)
17 ormap2_find::total=23629.2 usec, average=236.292 nsec(100000 samples)
```
ormap1 is single vector of key-value pairs, ormap2 is two separate vectors. For insertion, which is on critical path, two vector approach is faster. But generally speaking std map is the fastest.

[Uploading map_impl_test.zip…]()

Reply:
![image](https://github.com/rdiankov/openrave/assets/8113660/98aa92af-4a11-4e5d-8b2e-bf2dc1b1a80d)
Maybe having two vectors, we get better cache hit. 

With ormap2 with two separate vectors, we are accessing the name id vector most of the time, and we access the value vector just once after finding where to actually write value. We are reading small amount of memory contiguously. 

With vector of key-value pair, we are interested in keys only for most of the time, but keys are more sparsely located on memory and maybe higher chance of cache miss?

Reply:
Makes sense about the cache lines.